In [1]:
from roboflow import Roboflow
rf = Roboflow(api_key="OzhutvjcP3sHh1XS3Trj")
project = rf.workspace("titulacin").project("person-detection-9a6mk")
version = project.version(16)
dataset = version.download("yolov8")

loading Roboflow workspace...
loading Roboflow project...


In [9]:
import cv2
import numpy as np
import random
import tensorflow as tf
import matplotlib.pyplot as plt
from PIL import Image

In [1]:
from ultralytics import YOLO

# Load the model (choose a pre-trained model, e.g., 'yolov8n')
model = YOLO('yolov8n.pt')

# Train the model on your dataset
model.train(data=r'C:\Person-detection-16\data.yaml', epochs=10, imgsz=640, batch=16)

New https://pypi.org/project/ultralytics/8.3.63 available  Update with 'pip install -U ultralytics'
Ultralytics 8.3.60  Python-3.12.7 torch-2.5.1+cpu CPU (AMD Ryzen 7 3700U with Radeon Vega Mobile Gfx)
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=C:\Person-detection-16\data.yaml, epochs=10, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train16, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=Non

train: Scanning C:\Person-detection-16\train\labels.cache... 4407 images, 69 backgrounds, 0 corrupt: 100%|██████████| 4407/4407 [00:00<?, ?it/s]
val: Scanning C:\Person-detection-16\valid\labels.cache... 1071 images, 44 backgrounds, 0 corrupt: 100%|██████████| 1071/1071 [00:00<?, ?it/s]


Plotting labels to runs\detect\train16\labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
TensorBoard: model graph visualization added 
Image sizes 640 train, 640 val
Using 0 dataloader workers
Logging results to runs\detect\train16
Starting training for 10 epochs...
Closing dataloader mosaic

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       1/10         0G      1.581      2.271      1.601          6        640: 100%|██████████| 276/276 [47:43<00:00, 10.37s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [04:10<00:00,  7.37s/it]


                   all       1071       2293       0.47      0.431      0.408      0.177

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       2/10         0G      1.812       2.09      1.803         11        640: 100%|██████████| 276/276 [47:20<00:00, 10.29s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [04:11<00:00,  7.41s/it]


                   all       1071       2293      0.484      0.327      0.316      0.139

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       3/10         0G      1.791      1.956      1.805         10        640: 100%|██████████| 276/276 [55:55<00:00, 12.16s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [04:00<00:00,  7.07s/it]

                   all       1071       2293      0.606      0.467      0.491      0.232



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       4/10         0G      1.725      1.832      1.748         17        640: 100%|██████████| 276/276 [46:18<00:00, 10.07s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [04:48<00:00,  8.48s/it]

                   all       1071       2293      0.503      0.417      0.427      0.208



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       5/10         0G      1.634      1.653      1.668         10        640: 100%|██████████| 276/276 [1:02:59<00:00, 13.69s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [05:00<00:00,  8.85s/it]

                   all       1071       2293      0.709      0.538      0.619      0.336



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       6/10         0G      1.555      1.502      1.601         12        640: 100%|██████████| 276/276 [59:25<00:00, 12.92s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [05:09<00:00,  9.12s/it]

                   all       1071       2293      0.674       0.55      0.607      0.335



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       7/10         0G      1.484      1.405      1.535          9        640: 100%|██████████| 276/276 [1:02:47<00:00, 13.65s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [05:27<00:00,  9.64s/it]

                   all       1071       2293      0.745       0.58      0.683      0.389



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       8/10         0G      1.405      1.313      1.487         19        640: 100%|██████████| 276/276 [57:34<00:00, 12.52s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [04:59<00:00,  8.81s/it]

                   all       1071       2293      0.752      0.627      0.713      0.419



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


       9/10         0G      1.338      1.205       1.43         10        640: 100%|██████████| 276/276 [59:32<00:00, 12.94s/it] 
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [05:08<00:00,  9.08s/it]

                   all       1071       2293      0.795      0.656      0.748      0.461



      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size


      10/10         0G      1.278      1.122      1.388         11        640: 100%|██████████| 276/276 [59:31<00:00, 12.94s/it]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [05:20<00:00,  9.42s/it]

                   all       1071       2293      0.787      0.683      0.768      0.481



10 epochs completed in 10.136 hours.
Optimizer stripped from runs\detect\train16\weights\last.pt, 6.2MB
Optimizer stripped from runs\detect\train16\weights\best.pt, 6.2MB

Validating runs\detect\train16\weights\best.pt...
Ultralytics 8.3.60  Python-3.12.7 torch-2.5.1+cpu CPU (AMD Ryzen 7 3700U with Radeon Vega Mobile Gfx)
Model summary (fused): 168 layers, 3,005,843 parameters, 0 gradients, 8.1 GFLOPs


                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 34/34 [04:36<00:00,  8.12s/it]


                   all       1071       2293      0.787      0.683      0.768      0.481
Speed: 4.3ms preprocess, 232.1ms inference, 0.0ms loss, 1.4ms postprocess per image
Results saved to runs\detect\train16


ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x0000027721276810>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032,    0.033033,    0.034034,    0.035035,    0.036036,    0.037037,    0.038038,    0.039039,     0.04004,    0.041041,    0.042042,    0.043043,    0.044044,    0.045045,    0.046046,    0.047047,
          0.0480

In [2]:
results = model(r'C:\Person-detection-16\test\images', save=True, project='C:/Person-detection-16/test/images/Result', name='test_run', conf=0.3)


image 1/5 C:\Person-detection-16\test\images\photo-1634985492349-8589a9255cbe_jpg.rf.4f36728b2a976d7e05846712482d7c2f.jpg: 640x640 11 Personas, 215.0ms
image 2/5 C:\Person-detection-16\test\images\premium_photo-1661292128559-ec5206d3a05a_jpg.rf.1bd2bae0c61320527a41389ea21e225a.jpg: 640x640 1 Persona, 152.6ms
image 3/5 C:\Person-detection-16\test\images\premium_photo-1661430807726-7f3aca78277a_jpg.rf.f2054612302c3645a3bd3bff77871ea1.jpg: 640x640 1 Persona, 157.5ms
image 4/5 C:\Person-detection-16\test\images\premium_photo-1661483376273-ebce3f96a557_jpg.rf.97ce81f3227de3097209985d1a8b0510.jpg: 640x640 4 Personas, 152.5ms
image 5/5 C:\Person-detection-16\test\images\premium_photo-1661484857459-7256e961205d_jpg.rf.aa543481948b8207eea51c7f7b46471d.jpg: 640x640 1 Persona, 149.5ms
Speed: 6.2ms preprocess, 165.4ms inference, 1.4ms postprocess per image at shape (1, 3, 640, 640)
Results saved to C:\Person-detection-16\test\images\Result\test_run


In [3]:
metrics = model.val()
print(metrics)

Ultralytics 8.3.60  Python-3.12.7 torch-2.5.1+cpu CPU (AMD Ryzen 7 3700U with Radeon Vega Mobile Gfx)


val: Scanning C:\Person-detection-16\valid\labels.cache... 1071 images, 44 backgrounds, 0 corrupt: 100%|██████████| 1071/1071 [00:00<?, ?it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 67/67 [03:56<00:00,  3.52s/it]


                   all       1071       2293      0.787      0.683      0.768      0.481
Speed: 3.7ms preprocess, 197.0ms inference, 0.0ms loss, 1.1ms postprocess per image
Results saved to runs\detect\train162
ultralytics.utils.metrics.DetMetrics object with attributes:

ap_class_index: array([0])
box: ultralytics.utils.metrics.Metric object
confusion_matrix: <ultralytics.utils.metrics.ConfusionMatrix object at 0x0000027720198BC0>
curves: ['Precision-Recall(B)', 'F1-Confidence(B)', 'Precision-Confidence(B)', 'Recall-Confidence(B)']
curves_results: [[array([          0,    0.001001,    0.002002,    0.003003,    0.004004,    0.005005,    0.006006,    0.007007,    0.008008,    0.009009,     0.01001,    0.011011,    0.012012,    0.013013,    0.014014,    0.015015,    0.016016,    0.017017,    0.018018,    0.019019,     0.02002,    0.021021,    0.022022,    0.023023,
          0.024024,    0.025025,    0.026026,    0.027027,    0.028028,    0.029029,     0.03003,    0.031031,    0.032032, 

In [7]:
image_path = r'C:\Users\hafsa\anaconda3\envs\AI_Bootcamp\runs\detect\train16\results.png'
image = Image.open(image_path)
plt.imshow(image)
plt.axis("off")
plt.show()

<Figure size 640x480 with 1 Axes>

In [8]:

image_path = r'C:\Users\hafsa\anaconda3\envs\AI_Bootcamp\runs\detect\train16\confusion_matrix_normalized.png'
image = Image.open(image_path)
plt.imshow(image)
plt.axis("off")
plt.show()

<Figure size 640x480 with 1 Axes>

# Model Evaluation Summary:
- mAP@50: 76.8%
- mAP@50-95: 48.1%
- Precision: 78.7%
- Recall: 68.3%
- Inference Speed: 232ms per image (~4 FPS)

# Key Metrics:

Precision (metrics/precision(B)):
Value: 0.787
Indicates that 78.7% of the predictions made by the model are correct (low false positives).
Good: Precision above 0.75 is usually considered strong.

Recall (metrics/recall(B)):
Value: 0.683
Shows that the model correctly identifies 68.3% of the actual positive instances (low false negatives).
Moderate: While this is decent, improving recall could help the model detect more objects.

mAP50 (metrics/mAP50(B)):
Value: 0.768
Mean Average Precision at 50% IoU threshold indicates that the model is detecting objects well with reasonable localization accuracy.
Good: A value above 0.75 suggests effective performance at the baseline IoU threshold.

mAP50-95 (metrics/mAP50-95(B)):
Value: 0.481
Measures the model's performance across multiple IoU thresholds (50% to 95%), capturing stricter localization criteria.
Moderate: While typical, improvement in this metric can signify better object localization.

Fitness:
Value: 0.510
A composite score derived from the other metrics. It's used to compare different models or training runs.
Improving: A higher fitness value (>0.5) typically reflects balanced performance.

Speed:
Preprocessing: 4.27 ms
Inference: 232.12 ms
Postprocessing: 1.41 ms
These times are acceptable depending on your hardware and task requirements.